File: 746#Min_Cost_Climbing_Stairs.py
# 746. Min Cost Climbing Stairs

# O(n) space solution:
class Solution1:
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """
        if len(cost) < 3:
            return 0
        min_cost = [0] * len(cost)
        min_cost[0] = cost[0]
        min_cost[1] = cost[1]
        for i in range(2, len(cost)):
            min_cost[i] = min(min_cost[i-1], min_cost[i-2]) + cost[i]
        return min(min_cost[-1], min_cost[-2])

# O(1) space solution:
class Solution2(object):
    def minCostClimbingStairs(self, cost):
        f1 = f2 = 0
        for x in reversed(cost):
            f1, f2 = x + min(f1, f2), f1
        return min(f1, f2)

File: 746#MinCostClimbingStairs.go
// 746. Min Cost Climbing Stairs
package main

// O(1) space solution
func minCostClimbingStairs1(cost []int) int {
	f1 := 0
	f2 := 0
	for i := len(cost) - 1; i >= 0; i-- {
		f0 := cost[i] + min(f1, f2)
		f2 = f1
		f1 = f0
	}
	return min(f1, f2)
}

// Dynamic Programming
func minCostClimbingStairs2(cost []int) int {
	s := len(cost)
	minCost := make([]int, len(cost))
	minCost[0] = cost[0]
	minCost[1] = cost[1]
	for i := 2; i <= s-1; i++ {
		minCost[i] = min(minCost[i-1], minCost[i-2]) + cost[i]
	}
	return min(minCost[s-1], minCost[s-2])
}
